### PR TITLE
[Bigpicture] data replication services

### PR DESCRIPTION
**Pull request by jbygdell**
This PR adds the standalone data replication service (`sda-backup`) into the merged code base under the name `sync`
The `sync-api` originally comes from [here](https://github.com/neicnordic/sda-pipeline/pull/491), modified to be part of the merged code base. The functionality to send POST requests have been moved to the `sync` service.

- `sync`
  - Reads messages from the `mapping_stream`
  - Checks if the dataset is generated locally or remotely, remotely generated datasets are ignored.
  - Copies files from the archive, attaches and re-encrypts the header with the remote sites public key so that the complete files can be ingested as if they where uploaded by an end user.
  - Sends JSON data to the remote API host with information about all files in a dataset.
- `sync-api`
  - Parses the incoming JSON data and sends messages to ingest the files, assign file IDs and dataset ID

### CODE REVIEW

File: sda/cmd/syncapi/syncapi.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+			DecryptedChecksums: []schema.Checksums{{Type: "sha256", Value: files.ShaSum}},
+		}
+		finalizeMsg, err := json.Marshal(finalize)
+		if err != nil {
+			return fmt.Errorf("failed to marshal json messge: Reason %v", err)
+		}
+
+		if err := Conf.API.MQ.SendMessage(fmt.Sprintf("%v", time.Now().Unix()), Conf.Broker.Exchange, Conf.SyncAPI.AccessionRouting, finalizeMsg); err != nil {
+			return fmt.Errorf("failed to send mapping messge: Reason %v", err)
+		}
```
Comment (dbampalikis) [inline sda/cmd/syncapi/syncapi.go:195]:
```suggestion
		corrID := uuid.New().String()
		if err := Conf.API.MQ.SendMessage(corrID, Conf.Broker.Exchange, Conf.SyncAPI.IngestRouting, ingestMsg); err != nil {
			return fmt.Errorf("failed to send ingest messge: Reason %v", err)
		}

		accessionIDs = append(accessionIDs, files.FileID)
		finalize := schema.IngestionAccession{
			Type:               "accession",
			User:               blob.User,
			FilePath:           files.FilePath,
			AccessionID:        files.FileID,
			DecryptedChecksums: []schema.Checksums{{Type: "sha256", Value: files.ShaSum}},
		}
		finalizeMsg, err := json.Marshal(finalize)
		if err != nil {
			return fmt.Errorf("failed to marshal json messge: Reason %v", err)
		}

		if err := Conf.API.MQ.SendMessage(corrID, Conf.Broker.Exchange, Conf.SyncAPI.AccessionRouting, finalizeMsg); err != nil {
			return fmt.Errorf("failed to send mapping messge: Reason %v", err)
		}
```

File: sda/cmd/syncapi/syncapi.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/gorilla/mux"
```
Comment (dbampalikis) [inline sda/cmd/syncapi/syncapi.go:16]:
```suggestion
	"github.com/google/uuid"
	"github.com/gorilla/mux"
```

**Review by dbampalikis — CHANGES_REQUESTED** [review]

File: sda/cmd/syncapi/syncapi.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+		log.Infof("Web server is ready to receive connections at http://%s:%d", Conf.API.Host, Conf.API.Port)
+		if err := srv.ListenAndServe(); err != nil {
+			shutdown()
+			log.Fatalln(err)
+		}
+	}
+}
+
+func setup(config *config.Config) *http.Server {
+	r := mux.NewRouter().SkipClean(true)
```
[the remark this answers is not in the document]
Comment (jbygdell) [reply sda/cmd/syncapi/syncapi.go:76]: Then we need to do a full refactoring, better to save that for later.

**Review by dbampalikis — APPROVED** [review]
Looks good and with the latest fixes it works out of the box!

**Review by aaperis — APPROVED** [review]

**Pull request merged**